### ISSUE TITLE
LocalWorkspaceOptions.EnvironmentVariables should allow null values

### ISSUE THREAD

**Issue by MisinformedDNA** [pulumi/pulumi#6499]
I'm seeing nullability warnings since `LocalWorkspaceOptions.EnvironmentVariables` is of type `IDictionary<string, string>` and not `IDictionary<string, string?>`.

I would expect it to be the latter since it maintains consistency with CLR types:

[Environment.SetEnvironementVariable](https://docs.microsoft.com/en-us/dotnet/api/system.environment.setenvironmentvariable?view=net-5.0#System_Environment_SetEnvironmentVariable_System_String_System_String_) accepts a nullable string as its value and [Environment.GetEnvironmentVariable](https://docs.microsoft.com/en-us/dotnet/api/system.environment.getenvironmentvariable?view=net-5.0#System_Environment_GetEnvironmentVariable_System_String_) returns a nullable string.

## Steps to reproduce

```csharp
programArgs.EnvironmentVariables = new Dictionary<string, string>
{
    // possible null reference assignment
    ["ARM_CLIENT_ID"] = Environment.GetEnvironmentVariable("AzureServicePrincipalId"),
};
```

If you agree, I can file a PR.

## Affected feature

 area/automation-api

**Comment by justinvp** [comment]
Sounds reasonable.

> If you agree, I can file a PR.

That'd be great! Thanks!

**Comment by mikhailshilkov** [comment]
@MisinformedDNA We had to revert #6520 because it conflicted with #6654 that was merged several hours before it. In particular, `CliWrap` does not allow nullable environment values [here](https://github.com/pulumi/pulumi/blob/7f8985ab0fb870f0cd577f474255b348d6ebeefe/sdk/dotnet/Pulumi.Automation/Commands/LocalPulumiCmd.cs#L75). And this is likely because it [assigns](https://github.com/Tyrrrz/CliWrap/blob/0938c4dfb5d829181e387067ec62e74a2a72231f/CliWrap/Command.cs#L287-L288) those values to `System.Diagnostics.ProcessStartInfo.Environment` of type `IDictionary<string, string>`. We'd need to reconcile those somehow.

cc @t0yv0 who is the author of the CliWrap-based implementation.

**Comment by t0yv0** [comment]
Ouch! What is the desired meaning of null environment variable when staring as sub-process? AFAIK the OS does not admit nulls so we have a choice to either omit or set to default empty "". I can add a patch if we know what we want to do here. CC @MisinformedDNA

**Comment by mikhailshilkov** [comment]
@t0yv0 Do you know what https://docs.microsoft.com/en-us/dotnet/api/system.environment.setenvironmentvariable?view=net-5.0#System_Environment_SetEnvironmentVariable_System_String_System_String_ does?

**Comment by mikhailshilkov** [comment]
Ah, found the remark

> value is considered empty under any of the following conditions:
>   - It is null.
>   - It is String.Empty.

Should we standardize on non-nullable values then?

**Comment by t0yv0** [comment]
# Deletion

That remark is a great find. So SetEnvironmentVariable treats "" and null as the same, essentially it treats it as an instruction to "delete" the variable. If we want to match that in our handling of EnvironmentVariables we could do this actually:

1. If the user passes "FOO"=nil or "FOO"=empty, we invoke pulumi sub-process without FOO being defined; whether FOO is defined or not in the current environment

2. If the user does not pass "FOO", then we propagate "FOO" from the current environment to the pulumi sub-process

3. If the user passes "FOO"=bar (non-nil or empty), we use that value and propagate it to the pulumi sub-process

This would be a small patch to the code - I can do this if we want?

# Nullability

Separately, we need to decide whether to accept nulls in the API (and treat them same as empty values), or else to force the user to just pass empty values when they want deletion. I like the argument #6499 that System.Environment trades in nulls so there is precedent for allowing nulls.

**Comment by MisinformedDNA** [comment]
In my repro, I am (essentially) assigning one environment variable to another. So I think the API should not complain about a possible null.

Wouldn't [this](https://github.com/pulumi/pulumi/pull/6654/files#diff-ca2533e085f4365e71a44e2dd166bb28a79ea3610b41aff54b122b1764c5b12eR110) solve the issue?

```csharp
foreach (var pair in additionalEnv)
{
    env[pair.Key] = pair.Value ?? string.Empty;
}

**Comment by t0yv0** [comment]
Yeah or maybe like this: https://github.com/pulumi/pulumi/pull/6680 - what do you think?

**Comment by MisinformedDNA** [comment]
What if someone wants to remove a user/system environment variable? (I don't have a use case.) Removing it from our list will not remove it in the final var list. But assigning it null, actually will (or should).

**Comment by t0yv0** [comment]
Yeah let me experiment a bit to see what CliWrap does.

**Comment by MisinformedDNA** [comment]
Filed https://github.com/Tyrrrz/CliWrap/issues/109

**Comment by t0yv0** [comment]
Nice I commented there also. CliWrap does not ever remove env vars. It passes empty strings verbatim. And it actually treats the passed in env dict as "modifications" not the 'final reference set' of the sub-process environment. I am glad we checked. Let's wait on what CliWrap says before adding nulls in our API. In the meanwhile I can fix/simplify code not to pull from System.Environment as that is unnecessary.
